Samples/BackgroundBlur/BackgroundBlur/TransformBlur.cpp

```diff
@@ -0,0 +1,1550 @@
+        MFT_OUTPUT_STREAM_FIXED_SAMPLE_SIZE
+        // MFT_OUTPUT_STREAM_PROVIDES_SAMPLES
+        ;
+
+    if (m_pOutputType == NULL)
+    {
+        pStreamInfo->cbSize = 0;
+    }
+    else
+    {
+        pStreamInfo->cbSize = m_cbImageSize;
+    }
+
+    pStreamInfo->cbAlignment = 0;
+
+    return S_OK;
+}
+
+//-------------------------------------------------------------------
+// Name: GetAttributes
+// Returns the attributes for the MFT.
+//-------------------------------------------------------------------
+HRESULT TransformBlur::GetAttributes(IMFAttributes** pAttributes)
+{
+    HRESULT hr = MFCreateAttributes(pAttributes, 1);
+    hr = (*pAttributes)->SetUINT32(MF_SA_D3D_AWARE, FALSE);
+    return hr;
+}
+
+
+//-------------------------------------------------------------------
+// Name: GetInputStreamAttributes
+// Returns stream-level attributes for an input stream.
+//-------------------------------------------------------------------
+HRESULT TransformBlur::GetInputStreamAttributes(
+    DWORD           dwInputStreamID,
+    IMFAttributes** ppAttributes
+)
+{
+    // This MFT does not support any attributes, so the method is not implemented.
+    return E_NOTIMPL;
+}
+
+//-------------------------------------------------------------------
+// Name: GetOutputStreamAttributes
+// Returns stream-level attributes for an output stream.
+//-------------------------------------------------------------------
+HRESULT TransformBlur::GetOutputStreamAttributes(
+    DWORD           dwOutputStreamID,
+    IMFAttributes** ppAttributes
+)
+{
+    HRESULT hr = MFCreateAttributes(ppAttributes, 1);
+    hr = (*ppAttributes)->SetUINT32(MF_SA_MINIMUM_OUTPUT_SAMPLE_COUNT, 3);
+
+
+    return hr;
+}
+
+//-------------------------------------------------------------------
+// Name: DeleteInputStream
+//-------------------------------------------------------------------
+HRESULT TransformBlur::DeleteInputStream(DWORD dwStreamID)
+{
+    // This MFT has a fixed number of input streams, so the method is not implemented.
+    return E_NOTIMPL;
+}
+
+//-------------------------------------------------------------------
+// Name: AddInputStreams
+//-------------------------------------------------------------------
+HRESULT TransformBlur::AddInputStreams(
+    DWORD   cStreams,
+    DWORD* adwStreamIDs
+)
+{
+    // This MFT has a fixed number of output streams, so the method is not implemented.
+    return E_NOTIMPL;
+}
+
+//-------------------------------------------------------------------
+// Name: GetInputAvailableType
+// Description: Return a preferred input type.
+//-------------------------------------------------------------------
+
+HRESULT TransformBlur::GetInputAvailableType(
+    DWORD           dwInputStreamID,
+    DWORD           dwTypeIndex, // 0-based
+    IMFMediaType** ppType
+)
+{
+    TRACE((L"GetInputAvailableType (stream = %d, type index = %d)\n", dwInputStreamID, dwTypeIndex));
+
+    AutoLock lock(m_critSec);
```

Comment:
Which access is meant to be locked? It's better to restrict the lock and make clear what's shared here